jscomp/refmt/jsoo_refmt_main.ml

```diff
@@ -23,120 +23,780 @@
+        "type" , inject @@ Js.string "unexpected_error"
+      |])
+
+end
+
+(* One time setup for all relevant modules *)
+let () =
   Bs_conditional_initial.setup_env ();
-  Clflags.binary_annotations := false
+  Clflags.binary_annotations := false;
+  Misc.Color.setup (Some Always);
+  Lazy.force Super_main.setup;
+  Lazy.force Res_outcome_printer.setup
 
-let error_of_exn e =   
-  match Location.error_of_exn e with 
-  | Some (`Ok e) -> Some e 
+let error_of_exn e =
+  (match Location.error_of_exn e with
+  | Some (`Ok e) -> Some e
   | Some `Already_displayed
-  | None -> None
-
-let implementation ~use_super_errors impl str  : Js.Unsafe.obj =
-  let modulename = "Test" in
-  (* let env = !Toploop.toplevel_env in *)
-  (* Res_compmisc.init_path false; *)
-  (* let modulename = module_of_filename ppf sourcefile outputprefix in *)
-  (* Env.set_unit_name modulename; *)
-  Lam_compile_env.reset () ;
-  let env = Res_compmisc.initial_env() in (* Question ?? *)
-  (* let finalenv = ref Env.empty in *)
-  let types_signature = ref [] in
-  if use_super_errors then begin
-    Misc.Color.setup (Some Always);
-    Lazy.force Super_main.setup ;
-  end;
-
-  try
-    Js_config.jsx_version :=  3 ; (* default *)
-    let ast = impl (Lexing.from_string str) in     
-    let ast = Ppx_entry.rewrite_implementation ast in 
-    let typed_tree = 
-      let (a,b,_,signature) = Typemod.type_implementation_more modulename modulename modulename env ast in
-      (* finalenv := c ; *)
-      types_signature := signature;
-      (a,b) in      
-  typed_tree
-  |>  Translmod.transl_implementation modulename
-  |> (* Printlambda.lambda ppf *) (fun 
-    {Lambda.code = lam}
-
-    ->
-      let buffer = Buffer.create 1000 in
-      let () = Js_dump_program.pp_deps_program
-                          ~output_prefix:"" (* does not matter here *)
-                          NodeJS
-                          (Lam_compile_main.compile ""
-                              lam)
-                          (Ext_pp.from_buffer buffer) in
-      let v = Buffer.contents buffer in
-      Js.Unsafe.(obj [| "js_code", inject @@ Js.string v |]) )
-      (* Format.fprintf output_ppf {| { "js_code" : %S }|} v ) *)
-  with
-  | e ->
-      begin match error_of_exn e with
-      | Some error ->
-        Location.report_error Format.err_formatter  error;
-        Jsoo_common.mk_js_error error.loc error.msg
-      | None ->
-        let msg = Printexc.to_string e in
-        match e with
-        | Refmt_api.Migrate_parsetree.Def.Migration_error (_,loc)
-        | Refmt_api.Reason_errors.Reason_error (_,loc) ->
-          Jsoo_common.mk_js_error loc msg
-        | _ -> 
-          Js.Unsafe.(obj [|
-            "js_error_msg" , inject @@ Js.string msg;
-            "type" , inject @@ Js.string "error"
-          |])
-      end
+  | None -> None)
 
+module Converter = Refmt_api.Migrate_parsetree.Convert(Refmt_api.Migrate_parsetree.OCaml_404)(Refmt_api.Migrate_parsetree.OCaml_406)
```

Comment:
Why is this using the ocaml 4.04 ast ?

Reply:
I think this was a requirement of Reason's 3.6 api back then.